### PR TITLE
deployment-and-coverage

### PR DESCRIPTION
**Pull request by mikirov**
* Add hardhat project and deployment scripts
* Fix `join` method to accept validator address as argument
* Fix typos
* Add missing comments
* Remove `mutateState` modifier from SubnetActor contract
* Increase test coverage to 99.5%
* Pack storage variable slots

### CODE REVIEW

**Comment by mikirov** [comment]
Current dependencies on/for this PR:
* main
  * **PR #47** <a href="https://app.graphite.dev/github/pr/LimeChain/filecoin-ipc-actors-fevm/47" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/LimeChain/filecoin-ipc-actors-fevm/47?utm_source=stack-comment).

File: src/Gateway.sol [hunk 1]
```diff
@@ -20,6 +20,7 @@ import "openzeppelin-contracts/security/ReentrancyGuard.sol";
 import "openzeppelin-contracts/utils/structs/EnumerableSet.sol";
 import "openzeppelin-contracts/utils/structs/EnumerableMap.sol";
 import "openzeppelin-contracts/utils/Address.sol";
+import "forge-std/console.sol";
```
Comment (arrusev) [inline src/Gateway.sol:23]: we can remove the forge console import

**Review by arrusev — APPROVED** [review]
LGTM. Leaving 2 chores

File: src/SubnetActor.sol [hunk 2]
```diff
@@ -171,32 +149,38 @@ contract SubnetActor is ISubnetActor, ReentrancyGuard, Voting {
 
     receive() external payable onlyGateway {}
 
-    function join() external payable signableOnly notKilled mutateState {
-        if(msg.value == 0) revert CollateralIsZero();
-        
-        stake[msg.sender] += msg.value;
-        totalStake += msg.value;
+    function join(address validator) external payable signableOnly notKilled {
```
Comment (adlrocha) [inline src/SubnetActor.sol:152]: This is not an `address` in terms of crypto address  but a network address that needs to be persisted along with the validator information like we do in the rust actors: https://github.com/consensus-shipyard/ipc-actors/blob/afbc88798a8558954c3a7403cc1e0281b8ad0429/subnet-actor/src/types.rs#L27
```suggestion
    function join(string validator) external payable signableOnly notKilled {
```

**Review by adlrocha — CHANGES_REQUESTED** [review]
The PR is huge, it was quite hard to catch all of the details.  A few comments: 
- Can we add a README in the `scripts` folder explaining the scripts available and the instructions to deploy the contracts (or in the main repo README)?
- Also, the argument for `join` should be a network address, not a blockchain/crypto address

**Review by adlrocha — APPROVED** [review]
LGTM

**Pull request merged**